benchexec/container.py

```diff
@@ -834,6 +938,41 @@ def drop_capabilities(keep=[]):
         ctypes.byref(libc.CapHeader(version=libc.LINUX_CAPABILITY_VERSION_3, pid=0)),
         ctypes.byref(capdata),
     )
+    libc.prctl(libc.PR_CAP_AMBIENT, libc.PR_CAP_AMBIENT_CLEAR_ALL, 0, 0, 0)
+
+
+def cap_permitted_to_ambient():
+    """
+    Python version of util-linux/lib/caputils.c: cap_permitted_to_ambient()
+
+    Transfer all permitted capabilities to the inheritable set
+    and raise them in the ambient set if effective.
+    """
+    header = libc.CapHeader(libc.LINUX_CAPABILITY_VERSION_3, 0)
+    data = (libc.CapData * libc.LINUX_CAPABILITY_U32S_3)()
+
+    libc.capget(header, data)
+
+    data[0].inheritable = data[0].permitted
+    data[1].inheritable = data[1].permitted
+
+    libc.capset(header, data)
+
+    effective = (data[1].effective << 32) | data[0].effective
+    cap_last_cap = int(util.try_read_file("/proc/sys/kernel/cap_last_cap") or "0")
+    for cap in range(cap_last_cap + 1):
+        if effective & (1 << cap):
+            libc.prctl(libc.PR_CAP_AMBIENT, libc.PR_CAP_AMBIENT_RAISE, cap, 0, 0)
+
+
+def drop_ambient_cap():
+    """
+    Drop all ambient capabilities by removing them from the ambient set.
+    Corresponds to the opposite of cap_permitted_to_ambient.
```

Comment:
Hm: `cap_permitted_to_ambient` also transfers capabilities from `permitted` to `inheritable`. Do we need to undo this as well here?

Or can we skip changing `inheritable`?

Reply:
Looks like we cannot skip this. In order the make capabilities ambient, we first need to ensure that they are all inheritable. And it also seems that we cannot undo this here. If I modify `cap_permitted_to_ambient` like this:

```python
def cap_permitted_to_ambient():
    header = libc.CapHeader(libc.LINUX_CAPABILITY_VERSION_3, 0)
    data = (libc.CapData * libc.LINUX_CAPABILITY_U32S_3)()

    libc.capget(header, data)

    old_inheritable = [data[0].inheritable, data[1].inheritable]
    data[0].inheritable = data[0].permitted
    data[1].inheritable = data[1].permitted

    libc.capset(header, data)

    effective = (data[1].effective << 32) | data[0].effective
    cap_last_cap = int(util.try_read_file("/proc/sys/kernel/cap_last_cap") or "0")
    for cap in range(cap_last_cap + 1):
        if effective & (1 << cap):
            libc.prctl(libc.PR_CAP_AMBIENT, libc.PR_CAP_AMBIENT_RAISE, cap, 0, 0)

    data[0].inheritable = old_inheritable[0]
    data[1].inheritable = old_inheritable[1]

    libc.capset(header, data)
```

`fusermount3` will fail with "Operation not permitted".

Reply:
But we should certainly be able to undo it at the same place where we drop the ambient capabilities, shouldn't we?

Reply:
Yes, then we need to add a return value to `cap_permitted_to_ambient`, which would be `original_inheritable`, and pass this return value to `drop_ambient_cap`.

Reply:
Not if you inline everything into `permitted_cap_as_ambient`. After all, we do not need the other functions anymore.

Reply:
Sure, I understand.